google/calendar/server.py

```diff
@@ -0,0 +1,717 @@
+        raise ToolError(f"Failed to delete google calendar. HttpError: {err}")
+    except Exception as e:
+        raise ToolError(f"Failed to delete google calendar. Exception: {e}")
+
+
+@mcp.tool(
+    annotations={
+        "readOnlyHint": True,
+        "destructiveHint": False,
+    },
+    exclude_args=["cred_token"],
+)
+def list_events(calendar_id: Annotated[str, Field(description="calendar id")],
+                event_type: Annotated[Literal["birthday", "default", "focusTime", "fromGmail", "outOfOffice", "workingLocation"], Field(description="The type of event to list. Defaults to 'default'")] = "default",
+                single_event: Annotated[bool, Field(description="Whether to list only single event.")] = False,
+                time_min: Annotated[str | None, Field(description="Upper bound (exclusive) for an event's start time to filter by. if both time_min and time_max are None, time_min will be set to the current time. Must be an RFC3339 timestamp with mandatory time zone offset.")] = None,
+                time_max: Annotated[str | None, Field(description="Lower bound (exclusive) for an event's end time to filter by. Must be an RFC3339 timestamp with mandatory time zone offset.")] = None,
+                order_by: Annotated[Literal["updated"] | None, Field(description="Order by which to sort events. Valid options are: updated. If set, results will be returned in ascending order.")] = None,
+                q: Annotated[str | None, Field(description="Free text search terms to find events by")] = None,
+                max_results: Annotated[int, Field(description="Maximum number of events to return.", ge=1, le=500)] = 250,
+                cred_token: str = None) -> list:
+    """Lists events for a specific google calendar."""
+    if calendar_id == "":
+        raise ValueError("argument `calendar_id` can't be empty")
+    service = get_client(cred_token)
+
+    # optional parameters
+    params = {}
+    params["singleEvents"] = single_event
+    params["eventTypes"] = event_type
+    if time_min:
+        if not validate_rfc3339(time_min):
+            raise ValueError(
+                f"Invalid time_min: {time_min}. It must be a valid RFC 3339 formatted date/time string, for example, 2011-06-03T10:00:00-07:00, 2011-06-03T10:00:00Z"
+            )
+        params["timeMin"] = time_min
+    if time_max:
+        if not validate_rfc3339(time_max):
+            raise ValueError(
+                f"Invalid time_max: {time_max}. It must be a valid RFC 3339 formatted date/time string, for example, 2011-06-03T10:00:00-07:00, 2011-06-03T10:00:00Z"
+            )
+        params["timeMax"] = time_max
+
+    if (
+        not time_min and not time_max
+    ):  # if no time_min or time_max is provided, default to the current time, so it will list all upcoming events
+        time_min = get_current_time_rfc3339()
+        params["timeMin"] = time_min
+
+    if order_by:
+        params["orderBy"] = order_by
+
+    q = os.getenv("Q")
```

Comment:
Why are we reading an environment variable here?

Reply:
thanks lol, removed. somehow this line sneaks in